substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java

```diff
@@ -33,8 +33,9 @@
  */
 
 public class Range {
-    private static final String CLASS_DELIMITER = ".";
+    private static final String CLASS_DELIMITER = "::";
```

Comment:
@stooke are you OK with this change? If I recall correctly `.` was introduced to make debugging possible on Windows. Please let me know if we need to make this OS dependent.

Reply:
@zakkak n.b. the current DWARF code does not use this to write names in DIEs. That's is because class qualified method names are constructed by the debugger form the method root name and the name of the class it is declared in. Why do you need to insert a class-qualified method name into a DIE? Are your inline method DIEs not enclosed in a class DIE?

Reply:
I can always live with this change, but I don't understand why it's needed.  Because it's user-visible, (and because it doesn't look like Java(, my preference is that it doesn't change if it doesn't have to.

Reply:
The inline method DIEs are indeed enclosed in a class DIE, but not the class DIE of the class that defines them, but the class DIE that defines the corresponding method of a primary range. So if you have a compilation of method `A.foo` and foo happens to inline `D.bax` which is invoked through the also inlined `B.bar` and `C.baz`, i.e., the call chain is  `A.foo->B.bar->C.baz->D.bax` the DIEs doe `B.bar` and `C.baz` are generated in the class DIE of `A`.

#### Why? You may ask...
This is to avoid having to go through the source-mappings of all the compilations and "walking" the call-chain for each location in order to detect which types and methods we need to declare in the "Create all the types" phase https://github.com/oracle/graal/blob/ee82da58e48325981af1832ab33a46de2610b29a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java#L209-L221
The main problem with doing this is that the method declarations depend on information available in `Range` instances that are only created after the we create the types.

Reply:
> I can always live with this change, but I don't understand why it's needed. Because it's user-visible, (and because it doesn't look like Java(, my preference is that it doesn't change if it doesn't have to.

This is needed to ensure a consistent formatting in `backtrace`. Without this change backtraces appear like this (at least on linux):
```
#0  hello.Hello$Greeter::greeter(java.lang.String[] *) () at hello/Hello.java:34
#1  0x00000000005062b3 in hello.Hello::main(java.lang.String[] *) () at hello/Hello.java:67
#2  0x000000000041c325 in com.oracle.svm.core.JavaMainWrapper.runCore () at com/oracle/svm/core/JavaMainWrapper.java:146
#3  com.oracle.svm.core.JavaMainWrapper.run () at com/oracle/svm/core/JavaMainWrapper.java:182
#4  main () at com/oracle/svm/core/code/IsolateEnterStub.java:1
```
instead of like this:
```
#0  hello.Hello$Greeter::greeter(java.lang.String[] *) () at hello/Hello.java:34
#1  0x00000000005062b3 in hello.Hello::main(java.lang.String[] *) () at hello/Hello.java:67
#2  0x000000000041c325 in com.oracle.svm.core.JavaMainWrapper::runCore () at com/oracle/svm/core/JavaMainWrapper.java:146
#3  com.oracle.svm.core.JavaMainWrapper::run () at com/oracle/svm/core/JavaMainWrapper.java:182
#4  main () at com/oracle/svm/core/code/IsolateEnterStub.java:1
```
(see the difference in `#2` and `#3`).

Reply:
I remember there was some pushback from Oracle when a change was proposed, but I don't remember which way the proposal went at the time.  Just a warning that they may say something again.

My position has always been to never have the double colon in the first place, but I got over that a year ago.  I can more easily see it for class methods as opposed to instance methods - which your example doesn't have.  My solution at the time was to have '::' on Linux and '.' on Windows - it didn't fly.

I'm puzzled how the difference in the original traceback occurred - is it just a result of inlining?

Reply:
>Why? You may ask...
>
> This is to avoid having to go through the source-mappings of all the compilations and "walking" the call-chain for each location in order to detect which types and methods we need to declare in the "Create all the types" phase

Indeed. However, I think this is not what you ought to be doing. The issue here is that types reported by traversing the hosted universe  may not actually include a class for B, C or D. Even if the classes exist they may not, respectively, declare methods bar, baz or bax. I don't think it is a satisfactory solution to create dummy inline methods and shoehorn them into a class they don't actually belong to. For a start that will duplicate methods that have been independently inlined into different call trees (when A.foo and X.foo both inline B.bar etc -- n.b. this is very likely when B.bar is a Substrate method)

I recommend that you do add a step where you /complete/ the debuginfo type model by identifying and adding all the missing classes and methods and then use it to generate a complete suite of class and method DIEs which includes all the extra elements, nesting the methods in their owning class. *Even if* that requires an extra traversal of the ranges. That means the debugger can then be presented with a class DIE for every class it needs to report and a method DIE embedded in its owning class DIE for every method it needs to report. These DIEs  will clearly map to corresponding classes and methods in the input code base, whether or not they made it into the Hosted Universe as full classes or inlined remnants.

Of course the class entries and method entries you add during this extra step will not be able to be specified as fully as those reported from the Hosted Universe. For example, there will be no byte_size for the classes, no lo and hi addresses for the methods. However, that does not stop you generating partial DIEs for them.

This means you will need to adjust the class hierarchy under TypeEntry to allow for a full ClassEntry and a partial ClassEntry. Ditto for MethodEntry. And you will need to add abbrev section records that allow the partial records to be generated. This will be more code but it will implement a much clearer mapping from input Java types to debuginfo types to DIE records.

Reply:
Thanks for the feedback Andrew, I'll prepare a new revision.

Reply:
> The issue here is that types reported by traversing the hosted universe may not actually include a class for B, C or D.

I did some testing and it looks like the hosted universe includes the classes of inlined methods.

> I don't think it is a satisfactory solution to create dummy inline methods and shoehorn them into a class they don't actually belong to. For a start that will duplicate methods that have been independently inlined into different call trees (when A.foo and X.foo both inline B.bar etc -- n.b. this is very likely when B.bar is a Substrate method)

@adinn according to [DWARF 4](http://dwarfstd.org/doc/DWARF4.pdf) Section 3.3.8, the "dummy inline methods" are the so called "abstract instance entries" (Section 3.3.8.1), which even if we complete the debuginfo type model will need to be present, since "Each concrete inlined instance tree is uniquely associated with one (and only one) abstract instance tree. " (Section 3.3.8.2). As a result, the duplication that you mention is inevitable AFAIU. (what we can do better is reducing the duplication of abstract instance entries as described in https://github.com/oracle/graal/pull/2880#issuecomment-701096971.

Now the "trick" I currently do (as described in https://github.com/oracle/graal/pull/2880#issuecomment-800487184) is to use `DW_AT_name`, `DW_AT_decl_file` and `DW_AT_decl_line` to provide enough info for gdb to produce the backtrace instead of using `DW_AT_specification` which allows us to get away without a /complete/ debuginfo type model.

So a /complete/ debuginfo type model will indeed be cleaner, but will increase both the implementation complexity and the binary size (since it will include more method DIEs, it looks like the class DIEs are already there). Given the additional info please let me know if you still think it's worth revising the PR to have a /complete/ debuginfo type model?

Reply:
> I did some testing and it looks like the hosted universe includes the classes of inlined methods.

Yes, but are you sure of that? I am not (yet) convinced that this is going to be the case in all generated images.

> As a result, the duplication that you mention is inevitable AFAIU. (what we can do better is reducing the duplication of abstract instance entries as described in #2880 (comment).

Of course we have to include the referring concrete inlined instance tree info in the method that code is inlined into -- that's just how DWARF models inline code. I was talking about duplication of the abstract origin trees. I see that what you are proposing also finesses duplication of these abstract origin DIEs. However, I still don't buy your argument about increased size.

What you are suggesting is adding extra top-level subprogram DIEs for the missing methods and then having the inlined_subroutine records refer to them as the abstract origin. What I am suggesting is not very different. I recommend embedding the subprogram DIEs for the missing methods in class DIEs. In both cases the inlined_subroutine DIEs refer to one or the other as the abstract origin so nothing is changed there either way. So, how does my recommendation differ in practice?

  - if an inline owner class were omitted from the Hosted Universe my scheme would add a minimal class DIE for it
  - a minimal subprogram DIE embedded in its owner class DIE might need some fields not in your  top-level subprogram DIE

So, let's just note first off that if you are correct that all classes are present in the Hosted Universe then that first difference is not a difference. I still believe inlined classes may turn out to be missing (I think I came across such missing classes when debugging earlier versions of the code). Whether all classes are defined or not I still think we need the method to be attached to a class.  Without one the debugger does not consider the method to be a proper method associated with a C++ type. Inline methods will not be shown with other methods when you print the class type (that is true whether it is a real class or a dummy class). So, the code base will not be presented in a way that clearly reflects the source Java program.

How much effect does that second difference have? A minimal subprogram DIE in the class will probably include a few more attributes than than your top level DIE -- although we may be able to drop some of the attributes used for a real method. We don't have to provide child DIEs for stuff like parameter info (we don't know it). We *will* save space on the name since it only includes the method root name without the class qualifier. I don't believe this change is going to make a lot of difference to the overall DWARF info size -- I am sure whatever difference it does make will be a drop in the ocean compared to the amount of space used for the concrete inline method records.

Finally, I don't fine the space argument that compelling anyway. Debug info can be stripped out of an image and compressed so it does not really affect deployed image size. Even when unstripped it does not affect RSS (debuginfo is not mapped during normal execution). There is the issue of image build time. The extra debug I am suggesting info may take a bit longer to generate but the difference will be not be large compared to your alternative and will be way smaller relative to the whole image build time. And -- we still have a switch to disable inlining info if desired (which might be useful during development but I think would be foolish to use for a production build -- better to strip and save a larger deployed image debuginfo file in case debug turns out to be needed).

Reply:
Or I guess I could have just said:

  yes it's worth revising the PR to have a /complete/ debuginfo type model.

Reply:
Thanks for the extensive reply @adinn. Changing this back to draft till it's ready for review again. 

> I'm puzzled how the difference in the original traceback occurred - is it just a result of inlining?

@stooke yes a combination of inlining and the trick described in https://github.com/oracle/graal/pull/2880#issuecomment-800487184 which will go away so `::` won't need to be used as `CLASS_DELIMITER`, sorry for the noise.

Reply:
@zakkak No problem. Thank you for forcing me to clarify the requirements here to justify the extra work I am asking for.